### PR TITLE
Make Algae.Maybe.new(nil) return a Nothing

### PR DESCRIPTION
**Pull request by icidasset**
Was this what you had in mind?
Can also write some more tests if you want me to ☺️

### CODE REVIEW

**Comment by OvermindDL1** [comment]
Uh, a `Maybe.new(...)` returning anything but a valid non-Nothing value seems weird to me...  `nil` can be an entirely valid value to return, like the result of a computation that succeeded.

**Comment by expede** [comment]
@icidasset because `nil` can represent an actual value in Elixir, `Algae.Maybe.new(nil)` should return `Algae.Maybe.Just{just: nil}`. However, if you wanted to add a function like `Algae.Maybe.from_nillable/1` to distinguish `nil -> Nothing` for those that want it, that would work 👍

**Comment by icidasset** [comment]
Thanks @expede ! I've implemented your idea.

File: lib/algae/maybe.ex [hunk 1]
```diff
@@ -54,9 +54,43 @@ defmodule Algae.Maybe do
...
+      iex> new(9, nothing: 1)
+      %Algae.Maybe.Just{just: 9}
+
+  """
+  @spec new(any(), keyword()) :: Just.t()
+  def new(value, [nothing: nothing_value]) do
+    if value == nothing_value,
+      do: Nothing.new(),
+    else: Just.new(value)
+  end
```
Comment (expede) [inline lib/algae/maybe.ex]: Two very minor nitpicks:

1. The spec is missing a `Nothing`

```elixir
@spec new(any(), [nothing: any()]) :: Just.t() | Nothing.t()
```

(Also may as well add the expected key while we're in there ☝️)

2. Have you considered using pattern matching?

```elixir
def new(nope, [nothing: nope]), do: Nothing.new()
def new(value, _), do: Just.new(value)
```

**Comment by OvermindDL1** [comment]
Why not something more like this for `new/2`?
```elixir
@spec new(err, [nothing: err]) :: Nothing.t(err)
@spec new(v, [nothing: err]) :: Just.t(v) when v != err
```
And adjust `Just.t` and `Nothing.t` as appropriate as well?  (I think that is the right syntax for named typing, might need to make new types though?)

**Comment by expede** [comment]
@OvermindDL1 Yeah, the [automagic type generation](https://github.com/expede/algae/issues/26) could use some work. At the time (way back in the `.t` days), I was unaware that you could embed types inside types.

File: lib/algae/maybe.ex (hunk 1 above)
Comment (icidasset) [reply lib/algae/maybe.ex]: @expede Thanks! I didn't know you could match variables with pattern matching, neat! Spec is fixed as well.

**Comment by expede** [comment]
Looks awesome @icidasset — thanks!

**Pull request merged**